message_endpoint.go

```diff
@@ -0,0 +1,293 @@
+/*
+ * Copyright 2018 De-labtory
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package swim
+
+import (
+	"errors"
+
+	"github.com/DE-labtory/swim/pb"
+	"github.com/gogo/protobuf/proto"
+	"github.com/it-chain/iLogger"
+
+	"time"
+)
+
+var ErrSendTimeout = errors.New("Error send timeout")
+var ErrUnreachable = errors.New("Error this shouldn't reach")
+var ErrInvalidMessage = errors.New("Error invalid message")
+var ErrCallbackCollectIntervalNotSpecified = errors.New("Error callback collect interval should be specified")
+
+// sallback is called when target member sent back to local member a message
+// created field is for clean up the old callback
+type callback struct {
+	fn func(msg pb.Message)
+	created time.Time
+}
+
+// responseHandler manages callback functions
+type responseHandler struct {
+	callbacks map[string]callback
+	collectInterval time.Duration
+}
+
+func newResponseHandler(collectInterval time.Duration) *responseHandler {
+	h := &responseHandler{
+		callbacks: make(map[string]callback),
+		collectInterval: collectInterval,
+	}
+
+	go h.collectCallback()
+
+	return h
+}
+
+func (r *responseHandler) addCallback(seq string, cb callback) {
+	r.callbacks[seq] = cb
```

Comment:
@hihiboss
hihiboss 5 hours ago  Member
I wonder that sequence is string, not unsigned integer!

  @zeroFruit
zeroFruit 3 hours ago  Member
you can reference Seq in message.pb.go. and I think seq can be hash

  @hihiboss
hihiboss 2 hours ago  Member
Oh, I forgot it. You’re right! But I still wonder why seq is hash or string...
Plz explain that to me @junbeomlee ??